src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,58 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide the performance of ucol_getSortKey is worse when it is called with null output buffer
+            // the solution is to try to fill the sort key in a temporary buffer of size equal 4 x string length
+
+            int guessedSize = (int)Math.Min((long)source.Length * 4, 1024 * 1024); // 1MB is the biggest array that can be rented from ArrayPool.Shared without memory allocation
+
+            if (TryGetSortKey(source, options, guessedSize, out int actualSortKeyLength, out int hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash; // fast path, the buffer was big enough
+            }
 
-                byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
+            // the temporary buffer was too small but the actual sort key length was returned
```

Comment:
>/ the temporary buffer was too small but the actual sort key length was returned [](start = 13, length = 80)

I am not sure this is accurate. we need to try what ucol_getSortKey return when passing short buffer. I am guessing it will just return 0. I'll try this scenario by tomorrow or so.

Reply:
I just tried it and I m seeing it ucol_getSortKey return the expected length and not zero even the buffer is short. so, your code should be fine here.

---
In reply to: [289658069](https://github.com/dotnet/coreclr/pull/24889#discussion_r289658069) [](ancestors = 289658069)

Reply:
@tarekgh thanks!